lib/linked_list.rb

```diff
@@ -12,152 +13,176 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
-    end
-
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    # Time Complexity:  
-    # Space Complexity
-    def add_first(value)
-      raise NotImplementedError
-    end
-
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
-    def search(value)
-      raise NotImplementedError
-    end
-
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
-    def find_max
-      raise NotImplementedError
-    end
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
 
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
-    def find_min
-      raise NotImplementedError
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  # Time Complexity: O(1) constant because adding first node
+  # Space Complexity: O(1) constant because adding one node
+  def add_first(value)
+    new_node = Node.new(value)
+    if @head != nil
+      new_node.next = @head
     end
+    @head = new_node
+  end
 
+  # Additional Exercises
+  # returns the value in the first node
+  # returns nil if the list is empty
+  # Time Complexity: O(1) constant because only searching for value of head
+  # Space Complexity: O(1) constant because no additional space is needed
+  def get_first
+    return nil if !@head
+    return @head.data
+  end
 
-    # method that returns the length of the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def length
-      raise NotImplementedError
-    end
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n) where n equals the length of the linked lisst
+  # Space Complexity: O(n) where n equals the length of the linked list

```

Comment:
Space complexity O(n)?  You aren't creating any additional data structure or using recursion.